File: lecture_47/p4.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        int mp=-1;
        for(int i:piles)
            mp=max(i,mp);
        int lo=1;
        int hi=mp;
        int ans=0;
        while(lo<=hi)
        {
            int mid=lo+(hi-lo)/2;
            
            int rh=0;7827849856

            for(int i:piles)
            {
                rh+=ceil(i/(mid*1.0));
            }
            
            if(rh>h)
                lo=mid+1;
            else
            {
                ans=mid;
                hi=mid-1;
            }
        }
        return ans;
    }
};

File: lecture_48/p4.java
class Solution {
    public int[] kthSmallestPrimeFraction(int[] arr, int k) {
        double l=0;
        double h=1;
        while(true)
        {
            int p=0;
            int q=1;
            int n=arr.length;
            double mid=(l+h)/2.0;
            int count=0;
            int j=1;
            int temp=n-j;
            for(int i=0;i<n;i++)
            {
                while(j<n && arr[i] > mid*arr[j])
                {
                    j++;
                    temp--;
                }
                count+=temp;
                if(j<n && arr[i]*q > arr[j]*p)
                {
                    p=arr[i];
                    q=arr[j];
                }
            }
            if(count<k)
                l=mid;
            else if(count>k)
                h=mid;
            else
                return new int[]{p,q};
        }
    }
}



// alternative

class Solution {
public:
    vector<int> kthSmallestPrimeFraction(vector<int>& arr, int k) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        double l=0;
        double h=1;
        while(true)
        {
            int p=0;
            int q=1;
            int n=arr.size();
            double mid=(l+h)/2.0;
            int count=0;
            int j=1;
            int temp=n-j;
            for(int i=0;i<n;i++)
            {
                while(j<n && (arr[i]*1.0)/arr[j] > mid)
                {
                    j++;
                    temp--;
                }
                count+=temp;
                if(j<n && (arr[i]*1.0)/arr[j] > (p*1.0)/q)
                {
                    p=arr[i];
                    q=arr[j];
                }
            }
            if(count<k)
                l=mid;
            else if(count>k)
                h=mid;
            else
                return {p,q};
        }
    }
};